wgengine/packet/icmp.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package packet
+
+type ICMPType uint8
+
+const (
+	ICMPEchoReply    = 0x00
+	ICMPEchoRequest  = 0x08
+	ICMPUnreachable  = 0x03
+	ICMPTimeExceeded = 0x0b
+)
+
+func (t ICMPType) String() string {
+	switch t {
+	case ICMPEchoReply:
+		return "EchoReply"
+	case ICMPEchoRequest:
+		return "EchoRequest"
+	case ICMPUnreachable:
+		return "Unreachable"
+	case ICMPTimeExceeded:
+		return "TimeExceeded"
+	default:
+		return "Unknown"
+	}
+}
+
+type ICMPCode uint8
+
+const (
+	ICMPNoCode ICMPCode = 0
+)
+
+// ICMPHeader represents an ICMP packet header.
+type ICMPHeader struct {
+	IPHeader
+	Type ICMPType
+	Code ICMPCode
+}
+
+const (
+	icmpHeaderLength = 4
+	// icmpTotalHeaderLength is the length of all headers in a ICMP packet.
+	icmpAllHeadersLength = ipHeaderLength + icmpHeaderLength
+)
+
+func (h *ICMPHeader) Length() int {
```

Comment:
Was pointer receiver intentional? It could be a value receiver too. The struct is pretty small and doesn't contain pointers. (https://github.com/golang/go/wiki/CodeReviewComments#receiver-type)

Reply:
Thanks, changed everywhere except `ToResponse`, since that would clash with the `Header` interface  (would need to return a `Header` with all the associated headache).